(Raw)Autocomplete: Add optional [optionsViewOpenDirection] param

chrisbobbe opened the pull request:
Fixes #129748.

This follow's @gnprice's proposal in that issue's description:

> Specifically, I propose adding to `RawAutocomplete` and `Autocomplete` a field with a name like `alignment` whose type is an enum like
> 
> ```dart
> enum AutocompleteAlignment { above, below }
> ```
> 
> and that defaults to `AutocompleteAlignment.below`. Then in the future the dynamic version can be a third value for the enum.

except in this revision I emphasize "direction" over "alignment", to try to avoid appearing as though we can control how the options are aligned with a text field built outside `[RawAutocomplete.fieldViewBuilder]`. (In the case where callers don't pass `fieldViewBuilder`, they can use this new param for the direction, but they'll need to control the top/bottom alignment themselves, e.g., by putting the `RawAutocomplete` above or below their text field and parenting it so that its `SizedBox.shrink()` can have zero height.)

I've also added a mention in `fieldViewBuilder`'s dartdoc about what happens when it's null.

I've also updated the dartdoc on `focusNode` and `textEditingController` (via a template shared between those two) to reflect that you can omit `fieldViewBuilder`, which is a nicer API than passing a function that returns a `SizedBox.shrink()`.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

chrisbobbe committed:
RawAutocomplete: Add optional [optionsViewOpenDirection] param

Fixes #129748.

chrisbobbe committed:
RawAutocomplete: Update `split` doc template to reflect actual, nicer API

chrisbobbe committed:
RawAutocomplete: Comment in [fieldViewBuilder] on what happens when null

gnprice commented on packages/flutter/lib/src/widgets/autocomplete.dart:
```diff
@@ -150,6 +173,9 @@ class RawAutocomplete<T extends Object> extends StatefulWidget {
   /// Pass the provided [TextEditingController] to the field built here so that
   /// RawAutocomplete can listen for changes.
+  ///
+  /// If this parameter is null, then a [SizedBox.shrink] is built instead.
+  /// For how that pattern can be useful, see [textEditingController].
```
Cool, good to get this feature written down in the docs.

Oh hmm but this is inside the `{@template …}`/`{@endtemplate}` block. I think that means it'll show up on the [Autocomplete.fieldViewBuilder] doc as well. It doesn't belong there, right?

In that case I think a fix is to just put this paragraph after the `{@endtemplate}`, instead of before.

gnprice commented on packages/flutter/lib/src/widgets/autocomplete.dart:
```diff
@@ -161,9 +187,10 @@ class RawAutocomplete<T extends Object> extends StatefulWidget {
-  /// When following this pattern, [fieldViewBuilder] can return
-  /// `SizedBox.shrink()` so that nothing is drawn where the text field would
-  /// normally be. A separate text field can be created elsewhere, and a
+  /// When following this pattern, [fieldViewBuilder] can be omitted;
+  /// in that case, a `SizedBox.shrink()` is built instead,
+  /// so that a text field is not drawn where it would normally be.
```
Can leave the exact details of what omitting [fieldViewBuilder] does to that property's own doc:
```suggestion
  /// When following this pattern, [fieldViewBuilder] can be omitted,
  /// so that a text field is not drawn where it would normally be.
```

gnprice commented on packages/flutter/lib/src/widgets/autocomplete.dart:
```diff
@@ -421,7 +453,12 @@ class _RawAutocompleteState<T extends Object> extends State<RawAutocomplete<T>>
             link: _optionsLayerLink,
             showWhenUnlinked: false,
-            targetAnchor: Alignment.bottomLeft,
+            targetAnchor: widget.optionsViewOpenDirection == OptionsViewOpenDirection.up
+              ? Alignment.topLeft
+              : Alignment.bottomLeft,
```
Perhaps use a switch-expression — won't be much longer (perhaps one line longer for the close-brace), and it'll be helpful for giving the reader confidence that the conditions are exhaustive without having to look the type up.

gnprice commented on packages/flutter/lib/src/widgets/autocomplete.dart:
```diff
@@ -77,6 +77,28 @@ typedef AutocompleteFieldViewBuilder = Widget Function(
+  ///
+  /// The bottom edge of the options view will align with the top edge
+  /// of the text field built by [RawAutocomplete.fieldViewBuilder].
+  up,
+  /// Open downward.
```
nit:
```suggestion
  up,

  /// Open downward.
```

gnprice commented on packages/flutter/lib/src/widgets/autocomplete.dart:
```diff
@@ -77,6 +77,28 @@ typedef AutocompleteFieldViewBuilder = Widget Function(
+  /// Open downward.
+  ///
+  /// The top edge of the options view will align with the bottom edge
+  /// of the text field build by [RawAutocomplete.fieldViewBuilder].
```
nit:
```suggestion
  /// of the text field built by [RawAutocomplete.fieldViewBuilder].
```

gnprice commented on packages/flutter/test/widgets/autocomplete_test.dart:
```diff
@@ -421,6 +421,130 @@ void main() {
+                  fieldViewBuilder: (_, TextEditingController controller, FocusNode focusNode, __) {
+                    return TextField(controller: controller, focusNode: focusNode);
+                  },
+                  optionsViewBuilder: (_, __, ___) => const Placeholder())))));
+      await tester.showKeyboard(find.byType(TextField));
+      final Offset fieldBottomLeft = tester.getBottomLeft(find.byType(TextField));
+      final Offset optionsViewTopLeft = tester.getTopLeft(find.byType(Placeholder));
+      final Offset optionsViewCenter = tester.getCenter(find.byType(Placeholder));
+      expect(fieldBottomLeft, offsetMoreOrLessEquals(optionsViewTopLeft));
```
This test case is possibly a bit easier to think about if the options view looks a little more like a real options view would, showing the value of an option.

Also perhaps inline the finder calls into the expectations, since they're pretty concise and self-explanatory:
```suggestion
                  optionsViewBuilder: (_, __, ___) => const Text('a'))))));
      await tester.showKeyboard(find.byType(TextField));
      expect(tester.getBottomLeft(find.byType(TextField)),
        offsetMoreOrLessEquals(tester.getTopLeft(find.text('a'))));
```

gnprice commented on packages/flutter/test/widgets/autocomplete_test.dart:
```diff
@@ -421,6 +421,130 @@ void main() {
+      final Offset optionsViewTopLeft = tester.getTopLeft(find.byType(Placeholder));
+      final Offset optionsViewCenter = tester.getCenter(find.byType(Placeholder));
+      expect(fieldBottomLeft, offsetMoreOrLessEquals(optionsViewTopLeft));
+      expect(optionsViewCenter.dy, greaterThan(optionsViewTopLeft.dy));
```
This last check boils down to just saying that the options view doesn't have zero height — the center of any box will be below the top left of the same box by half the box's height.

I think it's OK to leave that check out. Particularly if the options view is something like a Text widget where it'd clearly be weird if it somehow got zero height.

gnprice commented on packages/flutter/test/widgets/autocomplete_test.dart:
```diff
@@ -421,6 +421,130 @@ void main() {
+    });
+
+    group('fieldViewBuilder not passed', () {
+      // Find the [SizedBox.shrink] that substitutes for a text field when
+      // [RawAutocomplete.fieldViewBuilder] is not passed.
+      final Finder sizedBoxShrinkFinder = find.descendant(
+        of: find.byType(RawAutocomplete<String>),
+        matching: find.byWidgetPredicate((Widget widget) {
+          return widget is SizedBox && widget.height == 0.0 && widget.width == 0.0;
```
Hmm, this feels a bit like it's probing into the internal details of the RawAutocomplete.

How about instead passing a key to the RawAutocomplete, like `key: autocompleteKey`, and then saying `find.byKey(autocompleteKey)`? I think that will cover what this is really getting at, which is that the options view is aligned in such-and-such a way against the box where the RawAutocomplete widget itself is laid out.

chrisbobbe replied:
The `SizedBox.shrink()` is explicitly part of `RawAutocomplete`'s API (as mentioned on [`.textEditingController`](https://api.flutter.dev/flutter/widgets/RawAutocomplete/textEditingController.html) and [`.focusNode`](https://api.flutter.dev/flutter/widgets/RawAutocomplete/focusNode.html) before this branch, and now in this branch on `fieldViewBuilder`), so I think this finder doesn't overstep into `RawAutocomplete`'s internal details.

But your alternative wouldn't be wrong, since I think it's assumed that the `RawAutocomplete` widget itself will be laid out to match the `SizedBox.shrink()`.

gnprice replied:
Yeah. It is part of the API, but it feels like it's a different part of the API from the one this test is about.

gnprice commented on packages/flutter/test/widgets/autocomplete_test.dart:
```diff
@@ -421,6 +421,130 @@ void main() {
+          MaterialApp(
+            home: Scaffold(
+              body: Column(
+                mainAxisAlignment: MainAxisAlignment.end,
```
What's the reason for using `MainAxisAlignment.end` in this test case? Combined with swapping the order in the Column, it seems like this will put the RawAutocomplete at the bottom, same as it is in the other test case.

chrisbobbe replied:
The `RawAutocomplete` will be positioned *near* the bottom of the screen, but I expect the text field to still be below it.

The goal is a reasonable layout, at least as far as making sure (some of) the options view is onscreen.

In this test, where we ask the options view to render upwards above the text field, accepting the default `MainAxisAlignment.start` would mean letting the options view render entirely offscreen, above the top of the screen. We might have avoided that in a different way, like adding a [`Scaffold.appBar`](https://api.flutter.dev/flutter/material/Scaffold/appBar.html), but this way seemed fine.

gnprice replied:
Ahh, I see — I was confusing the effect of this parameter with that of [Column.verticalDirection](https://api.flutter.dev/flutter/widgets/Flex/verticalDirection.html).

That reasoning makes sense. Rereading this test with the right understanding of that parameter in mind, I see that this and the other test are symmetrical just as one would hope.

gnprice reviewed (commented):
Thanks! Generally this looks great — I like the more specific name for the enum and the parameter. Comments below.

chrisbobbe committed:
RawAutocomplete: (response to gnprice's code review)

This responds to gnprice's review at
  https://github.com/flutter/flutter/pull/129802#pullrequestreview-1506268242

chrisbobbe committed:
RawAutocomplete: Update optionsViewBuilder to mention new param

chrisbobbe commented:
Thanks for the review! Added https://github.com/flutter/flutter/pull/129802/commits/003f88a18df5a1af969123f5d082ac89827d6a88 with corresponding changes.

I've also added https://github.com/flutter/flutter/pull/129802/commits/ba7d49da613a578a9937758a6ad7370bae6abb69 on top of that; it's something I forgot about in my first draft.

Oh! Also: this only touches `RawAutocomplete`, but https://github.com/flutter/flutter/issues/129748 covers `Autocomplete` too. Should I mark this as a draft until I update `Autocomplete` too (which I can aim to do soon)?

gnprice reviewed (approved):
Thanks! This now all LGTM.

> Oh! Also: this only touches `RawAutocomplete`, but #129748 covers `Autocomplete` too. Should I mark this as a draft until I update it to add the param on `Autocomplete` too (which I can aim to do soon)?

Ah, good thought. I'll mark as draft (it looks like I have a button to do that too), and please go ahead and re-mark as ready once you add that bit.

chrisbobbe committed:
Autocomplete: Add optionsViewOpenDirection, passing it to RawAutocomplete

chrisbobbe committed:
autocomplete: Add "Autocomplete" prefix to OptionsViewOpenDirection

Like similar types in widgets/autocomplete.dart, such as:

  AutocompleteOptionsBuilder
  AutocompleteOnSelected
  AutocompleteOptionsViewBuilder

and so on.

It makes for a pretty long name, though; hmm.

chrisbobbe commented:
OK! Pushed a commit that makes this new param available on `Autocomplete`, and also another commit with a rename for the enum to follow a pattern; PTAL.

gnprice commented on packages/flutter/lib/src/material/autocomplete.dart:
```diff
@@ -90,6 +91,9 @@ class Autocomplete<T extends Object> extends StatelessWidget {
   /// default.
   final AutocompleteOptionsViewBuilder<T>? optionsViewBuilder;
 
+  /// {@macro flutter.widgets.RawAutocomplete.fieldViewBuilder}
+  final AutocompleteOptionsViewOpenDirection optionsViewOpenDirection;
```
```suggestion
  /// {@macro flutter.widgets.RawAutocomplete.optionsViewOpenDirection}
  final AutocompleteOptionsViewOpenDirection optionsViewOpenDirection;
```

chrisbobbe replied:
Eep, thanks!

gnprice commented on packages/flutter/test/widgets/autocomplete_test.dart:
```diff
@@ -421,6 +421,109 @@ void main() {
+          home: Scaffold(
+            body: Center(
+              child: RawAutocomplete<String>(
+                optionsViewOpenDirection: AutocompleteOptionsViewOpenDirection.up,
```
Hmm yeah, with the "Autocomplete" prefix on the name of the enum, this gets kind of long.

I think it's OK to leave that prefix off, and make it just `OptionsViewOpenDirection`. My reasoning is:

* Those other types around it that have that prefix are typedefs, not enums. So there's basically no cost in the niceness of the API from making their names long, because calling code won't typically mention them.

* On the other side, the risk from having a less specific name is that it effectively lives in a single shared namespace with the other public names in the framework, and so it might collide in the future with some name we want elsewhere, or sound confusingly like something that'd appear elsewhere in the framework.

  But I think "options view open direction" is already specific enough that that's not a big worry: if there's something else that might be thought of as an "options view" and it has an "open direction", then I think it'd have to have enough in common with `RawAutocomplete` that basically the same set of values for the "open direction" will make sense for that other thing as for `RawAutocomplete`. Namely the "options view" can go above, or go below, or better yet in the future would be able to automatically go where it fits but the app should decide whether to prefer above or below.

  So if there ever is such a collision, I think we can cheerfully use the same enum for both. We'd just revise its docs to be worded more generally, and to mention both uses under "See also:".

chrisbobbe replied:
That reasoning makes sense; will do.

gnprice reviewed (commented):
Thanks! Small comments.

chrisbobbe committed:
autocomplete: (response to gnprice's code review)

This responds to gnprice's review at
  https://github.com/flutter/flutter/pull/129802#pullrequestreview-1507779331

chrisbobbe commented:
Thanks for the review! Added https://github.com/flutter/flutter/pull/129802/commits/3f9a61ef9d1e2775ce33fe99de6c75e0d608ed92.

gnprice reviewed (approved):
LGTM! Thanks for adding this.

I'll see about trying to find a second reviewer.

TheCarpetMerchant commented:
It would be much more useful to get an option for the widget to calculate which direction it should show the list in by itself, like external packages like flutter_typeahead already do.

gnprice commented:
> It would be much more useful to get an option for the widget to calculate which direction it should show the list in by itself, like external packages like flutter_typeahead already do.

That would be useful too, but is more complex. There are plenty of use cases where the app knows it always wants the options to go above (or always below), so this option is already useful in itself.

See discussion in the issue #129748, and the related issues linked from there. For the feature you mentioned, I think the issue to track is #101620.

Even in a future where the widget can dynamically decide whether to go above or below, we'll still want an option like this, because there's still the question of which direction it should prefer (which direction it should go in when there's room in both directions). The API in this PR is designed so that it can smoothly be extended for that future, using an enum.

justinmc commented on packages/flutter/test/material/autocomplete_test.dart:
```diff
@@ -507,4 +507,41 @@ void main() {
     checkOptionHighlight(tester, 'northern white rhinoceros', null);
   });
+
+  group('optionsViewOpenDirection', () {
+    testWidgets('default (down)', (WidgetTester tester) async {
+      await tester.pumpWidget(
+        MaterialApp(
+          home: Scaffold(
+            body: Autocomplete<String>(
+              optionsBuilder: (_) => <String>['a']))));
```
The `_` parameter should instead give the full type and name according to the styleguide: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-anonymous-parameter-names

justinmc replied:
Also see the position of your closing parentheses related to this: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#if-you-have-a-newline-after-some-opening-punctuation-match-it-on-the-closing-punctuation

justinmc reviewed (commented):
Some formatting to clean up, but otherwise this looks good.  Also see my comment about the format of optionsViewOpenDirection.  I think it's probably fine as-is but wanted to put my thoughts down.

justinmc commented:
The `Windows build` test that's failing looks like an infrastructure flake to me.  If it doesn't go green, try pushing a merge commit.

chrisbobbe committed:
Merge remote-tracking branch 'upstream/main'

chrisbobbe committed:
autocomplete: (response to justinmc's code review)

This responds to:
  https://github.com/flutter/flutter/pull/129802#pullrequestreview-1507943594

chrisbobbe commented:
Thanks for the review! Fixes pushed; PTAL. 🙂

justinmc commented on packages/flutter/test/widgets/autocomplete_test.dart:
```diff
@@ -421,6 +421,144 @@ void main() {
+        ),
+      );
+      await tester.showKeyboard(find.byType(TextField));
+      expect(tester.getBottomLeft(find.byType(TextField)),
+        offsetMoreOrLessEquals(tester.getTopLeft(find.text('a'))));
```
Nit: Another place where the formatting is off from the styleguide standard.  Here and in the other tests in this file.

```suggestion
      expect(
        tester.getBottomLeft(find.byType(TextField)),
        offsetMoreOrLessEquals(tester.getTopLeft(find.text('a'))),
      );
```

chrisbobbe replied:
I'm happy to change it if you prefer, but I think in this case, this more compact form is actually consistent with the style guide. Do you have the same section in mind as last time ([this one](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#if-you-have-a-newline-after-some-opening-punctuation-match-it-on-the-closing-punctuation))? It says:

> If you have a newline after some opening punctuation, match it on the closing punctuation.

Note "newline after [the] opening punctuation". In this case, my `expect(` isn't followed by a newline. There's a newline between the two arguments, but that doesn't force adding newlines near the `(` and `)`, does it? Under the "GOOD" heading, there's a case like the code in this revision:

```dart
// BAD:
  foo(
    bar, baz);
  foo(
    bar,
    baz);
  foo(bar,
    baz
  );

// GOOD:
  foo(bar, baz);
  foo(
    bar,
    baz,
  );
  foo(bar,
    baz);
```

in particular

```dart
  foo(bar,
    baz);
```

. In the code here, in place of `foo`, `bar`, and `baz`, we have `expect`, `tester.getBottomLeft(find.byType(TextField))`, and `offsetMoreOrLessEquals(tester.getTopLeft(find.text('a')))`, respectively.

justinmc replied:
Ah sorry you're right, that's explicitly allowed!

justinmc reviewed (approved):
LGTM :+1:
